src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -6426,16 +6426,126 @@ internal Ast CommandRule(bool forDynamicKeyword)
 
         #region Expressions
 
-        private ExpressionAst ExpressionRule()
+        private ExpressionAst ExpressionRule(bool endNumberOnTernaryOpChars = false)
         {
             // G  expression:
             // G      logical-expression
             // G
             // G  logical-expression:
```

Comment:
Hmm, I'm not sure enough the updated grammar here is accurate. Do you think it's fine to just put `expression: ternary-expression`?

Reply:
I think if `ExpressionRule()` represents all possible expressions, each kind of expression should be a possible expansion of `expression`.

Ideally we shouldn't have any grammar descriptions that look like `rule: another_rule`, since the LHS then has no purpose and is redundant; the RHS of a grammatical rule description should be non-trivial.

In this case though, I think `logical-expression` was supposed to mean a logical boolean expression. Given that the expression rule can recognise more than that, that should probably be dispensed with. But also the `ExpressionRule` can recognise unary expressions, which aren't accounted for in the grammar comment.

That might lead to something like:

```
expression:
    unary-expression
    binary-expression
    ternary-expression
```

That might be suitable, but not a very good description of either how the grammar or the implementation works. In something like `StatementRule()` that's the description, but for statement recognition there's a simple one-token lookahead (for a keyword or lack thereof).

Here we read a whole unary expression (I presume) and then look ahead for an operator and decide based on the operator whether it's binary or ternary. Other expression rules describe things like this using quite detailed unrollings of the grammar that list the literals they look ahead for and even cascading grammar rules to describe precedence and associativity (compare [this](https://github.com/PowerShell/PowerShell/blob/5ef258ef68952be81f2a65345574f330934ae9ef/src/System.Management.Automation/engine/parser/Parser.cs#L6434-L6459) to [this](http://www.cs.ecu.edu/karl/5220/spr16/Notes/CFG/precedence.html)).

Bottom line, my thinking is:
- The grammar statement at the top of the rule should match the granularity of the implementation in the given parser rule
- The RHS of any grammar rule expansion should be non-trivial (i.e. should expand to either a sequence of non-terminals or a construction containing a terminal)
- Given the effort gone to in existing grammar comments to unroll the grammar properly, we should try and keep the same as things mutate

Reply:
Ah, I see that `binary-expression` implicitly encompasses `array-literal-expression` which in turn encompasses `unary-expression`. I suppose that's just the way things are, so the current grammar comment looks about right to me (minus the `logical-expression`)